### PR TITLE
Feat/add moss tts

### LINKED ISSUE #1319: [New Model]: OpenMOSS-Team/MOSS-TTS
### The model to consider.

[OpenMOSS-Team/MOSS-TTS](https://huggingface.co/OpenMOSS-Team/MOSS-TTS)

### The closest model vllm-omni already supports.

_No response_

### What's your difficulty of supporting the model you want?

_No response_

### Use case and motivation

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by anurag12-webster**
## Purpose

Add MOSS-TTS (MossTTSDelay) support to vLLM-Omni.

This PR resolves [#1319](https://github.com/vllm-project/vllm-omni/issues/1319) by integrating MOSS-TTS as a **single-stage MVP TTS model** and wiring it into model registry, stage config, docs, examples, and tests.

Key implementation points:
- Added MOSS-TTS integration under `vllm_omni/model_executor/models/moss_tts/`
- Registered MOSS-TTS config/architecture in framework registry paths
- Added `moss_tts.yaml` stage config
- Added offline inference example + docs + supported models update
- Added offline e2e tests for basic MOSS-TTS flows

Important framework fix:
- Updated `arg_utils.py` so local `MossTTSDelayConfig` is used during model config creation (temporarily setting `trust_remote_code=False`, then restoring it), avoiding remote-config mismatch and runtime attribute errors.

RFC alignment note:
- This PR is intentionally MVP-first and does **not** implement full RFC #1225 async-chunk architecture yet.
- Follow-up work will add 2-stage AR -> AudioDecoder flow, stage input processor, and inter-stage streaming.

## Test Plan

Environment notes:
- `vllm-omni` and `vllm` are installed separately in this setup
- Initial run on RTX 4090 hit OOM
- Main validation was completed on A100 GPU

Dependency issue found during testing:
- `vllm` installation pulled `transformers==4.57.0`, which did not include required MOSS-TTS components in this environment
- Upgraded `transformers` to `5.1.0` for testing
image

- Updated `vllm` from `0.15.0` to `0.15.1` in the same environment

Commands/scripts used:
1. Offline inference
   - `python examples/offline_inference/moss_tts/end2end.py`
2. Offline e2e tests
   - `pytest tests/e2e/offline_inference/test_moss_tts.py -v`
3. Manual quality checks
   - English TTS sample generation
   - Chinese TTS sample generation
   - Voice cloning sample generation

Audio samples:
- English: [output_0_e3714fda-1dcd-44d8-bcb6-656b8af8f611.wav](https://github.com/user-attachments/files/25284263/output_0_e3714fda-1dcd-44d8-bcb6-656b8af8f611.wav)
- Chinese: [output_0_8cef6e4a-66ba-4603-8471-dfd9dd006367.wav](https://github.com/user-attachments/files/25284271/output_0_8cef6e4a-66ba-4603-8471-dfd9dd006367.wav)
- Cloned from a reference clip: [voice_clone_test.wav](https://github.com/user-attachments/files/25284284/voice_clone_test.wav)

## Test Result

- RTX 4090: OOM during model test
- A100: tests and inference completed successfully after dependency updates
- Verified successful generation for:
  - English sample audio
  - Chinese sample audio
  - Voice cloning sample audio
- Offline e2e test passes on A100 (`tests/e2e/offline_inference/test_moss_tts.py`)

### Checklist

- [x] Purpose included, with issue linkage
- [x] Test plan included with scripts/commands and env notes
- [x] Test results included (including hardware/dependency behavior)
- [x] Documentation updates included (`supported_models.md`, offline inference docs/examples)
- [ ] Release notes update (optional; can be added if maintainers request this user-facing change listed)

### CODE REVIEW

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+                torch.arange(n_vq, dtype=int, device=device).expand(batch_size, n_vq) > delayed_lengths.unsqueeze(1) - 1
+            )
+            post_audio_mask[delayed_lengths == torch_int64_max] = True
+            sampling_audio_mask = pre_audio_mask & post_audio_mask
+            next_audio_tokens[~sampling_audio_mask] = self.config.audio_pad_code
+
+            if sampling_audio_mask.sum() > 0:
+                audio_logits = torch.stack(next_token_logits[1:], dim=1)[
+                    sampling_audio_mask
+                ]  # torch.stack -> [batch_size, n_vq - 1, vocab_size]
```
[the remark this answers is not in the document]
Comment (anurag12-webster) [reply vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py:524]: this.. is upstream model code from OpenMOS.. not introduced by this PR.

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+        try:
+            import flash_attn  # noqa: F401
+
+            attn_kwargs = {"attn_implementation": "flash_attention_2"}
+        except ImportError:
+            logger.warning("Flash-Attn is not installed. Using default PyTorch attention implementation.")
+            attn_kwargs = {}
+
+        self.model = MossTTSModel.from_pretrained(
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:36]: Model loaded via HF `from_pretrained`, bypassing vLLM's weight loader. `load_weights()` (line 202) is a no-op that just marks params as loaded. This means double loading and no benefit from vLLM's sharded/safetensor optimizations. Consider following the `qwen3_tts` pattern.

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+            input_ids: Input token IDs
+            multimodal_embeddings: Optional multimodal embeddings (not used for TTS)
+            is_multimodal: Optional mask indicating multimodal tokens (not used for TTS)
+            **kwargs: Additional arguments
+
+        Returns:
+            torch.Tensor: Embedded representations of input_ids
+        """
+        return torch.zeros(
+            (input_ids.shape[0], input_ids.shape[1], 1024),
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:182]: Hidden dim hardcoded as `1024`, but the actual Qwen3-8B backbone has `hidden_size=4096`. Should derive from config.

File: vllm_omni/engine/arg_utils.py [hunk 4]
```diff
@@ -130,6 +142,11 @@ def create_model_config(self) -> OmniModelConfig:
         }
         stage_connector_config["extra"]["stage_id"] = self.stage_id
 
+        # Force local HF config classes only for MOSS-TTS to avoid remote-config
+        # mismatches while preserving existing trust_remote_code behavior for other models.
+        saved_trust = self.trust_remote_code
+        config_trust_remote_code = False if self.model_arch == "MossTTSDelayModel" else self.trust_remote_code
```
Comment (linyueqian) [inline vllm_omni/engine/arg_utils.py:148]: This `trust_remote_code` save/restore hack is duplicated in both `create_model_config` methods (here and ~line 310). Also fragile - if local config registration fails (which it does on transformers 4.x), the `False` override causes a different failure cascade.

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+            **kwargs:
+                Forwarded to AutoModel.from_pretrained(...).
+
+        Returns:
+            MossTTSModel: Wrapper instance containing model, processor.
+        """
+        AutoConfig.register("moss_tts_delay", MossTTSDelayConfig)
+        AutoModel.register(MossTTSDelayConfig, MossTTSDelayModel)
+
+        codec_path = kwargs.pop("codec_path", "OpenMOSS-Team/MOSS-Audio-Tokenizer")
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:274]: `codec_path` hardcoded with no way to override from CLI or stage config. Users with local checkpoints have to modify source code.

**Review by linyueqian — CHANGES_REQUESTED** [review]
Tried running this end-to-end on H200 with current `main` (`vllm==0.16.0`, `transformers==4.57.3`). **Does not work** - requires `transformers>=5.0`.

Specific breakages:
- `modeling_moss_tts.py:22` - `from transformers import initialization as init` (5.x only)
- `processing_moss_tts.py:24` - `processing_utils.MODALITY_TO_BASE_CLASS_MAPPING` (5.x only)
- Upstream `OpenMOSS-Team/MOSS-Audio-Tokenizer` HF repo uses `PreTrainedConfig` (capital T, renamed in 5.x)

@hsliuustc0106 Is there a plan to upgrade to `transformers` v5? Also, is there a reason this PR needs `transformers` directly rather than going through vLLM's built-in model loading?

Also, the OpenMOSS family has multiple model variants (MOSS-TTS, MOSS-TTSD, MOSS-VoiceGenerator, MOSS-SoundEffect, MOSS-TTS-Realtime) per https://huggingface.co/OpenMOSS-Team/MOSS-TTS. Are there plans to support the other variants as well? If so, it might be worth designing the integration with that in mind rather than coupling everything to the single delay-pattern architecture.

**Comment by anurag12-webster** [comment]
hey thanks.. @linyueqian let me resolve these issue and restructure the PR in order to integrate full family of of MOSS-TTS models.

**Comment by anurag12-webster** [comment]
I’ll update this PR to be transformers 4.57.x compatible (remove 5.x-only APIs), keep proper vLLM weight loading (no eager HF main-model load), and scope this PR to MossTTSDelayModel with follow-up support for other MOSS variants.

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+# limitations under the License.
+"""Modeling classes for MossTTSDelay."""
+
+from dataclasses import dataclass
+
+import torch
+import torch.nn as nn
+from torch.nn import CrossEntropyLoss
+from tqdm import tqdm
+from transformers import initialization as init
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py:22]: I believe `from transformers import initialization as init` only exists in transformers 5.x. Since vllm currently pins to 4.x, this would likely fail at import time. Could you verify which transformers version this targets?

File: vllm_omni/model_executor/models/moss_tts/processing_moss_tts.py [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+import os
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, cast
+
+import torchaudio
+from transformers import processing_utils
+
+processing_utils.MODALITY_TO_BASE_CLASS_MAPPING["audio_tokenizer"] = "PreTrainedModel"
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/processing_moss_tts.py:24]: This mutates a global dict on `processing_utils` at import time, and it is also a 5.x-only attribute. Even after fixing the version issue, monkey-patching a library internal mapping as a module-level side effect is fragile -- if transformers changes the internal structure this breaks silently. Is there a way to register this through the official API instead?

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py (hunk 2 above)
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:36]: Echoing @linyueqian point here. Loading the full model via `MossTTSModel.from_pretrained` in the constructor means vLLM weight distribution, sharding, and memory management are completely bypassed. `load_weights` below (line 202) just records param names without doing anything. For a model this size (8B), this matters a lot -- you lose tensor-parallel support and get doubled peak memory during init. Is there something about the MOSS-TTS architecture that prevents using vLLM standard weight loading path?

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py (hunk 3 above)
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:182]: The value `1024` appears to be hardcoded here, but Qwen3-8B has `hidden_size=4096`. I think this might produce wrong-shaped tensors — could you double-check whether this should be derived from the model config instead?

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+    add_start_docstrings_to_model_forward,
+    logging,
+    replace_return_docstrings,
+)
+
+from .configuration_moss_tts import MossTTSDelayConfig
+from .inference_utils import find_last_equal_C, sample_token
+
+try:
+    from .processing_moss_tts import AssistantMessage, MossTTSDelayProcessor, UserMessage
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py:38]: Bare `except Exception` here means any real import error (missing dependency, syntax error in processing_moss_tts.py, etc.) gets silently swallowed and you get `None` values that blow up later with a confusing traceback. Could you narrow this to `ImportError` at minimum?

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+
+        # Three stages: 1. non-audio; 2. audio without delay; 3. audio with delay
+        audio_lengths = torch.zeros(batch_size, dtype=torch.int64, device=device)  # A value of 0 indicates stage 1.
+        torch_int64_max = torch.iinfo(torch.int64).max
+        delayed_lengths = torch.full(
+            (batch_size,), torch_int64_max, dtype=torch.int64, device=device
+        )  # The maximum value indicates stage 2.
+
+        # Consider the case where audio_start is already present in input_ids during continuation.
+        # NOTE: Currently we do not consider any case where the input has already entered the delay stage.
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/modeling_moss_tts.py:432]: `tqdm` progress bar inside a library `generate()` method will be noisy when running under a server. Worth gating behind a `verbose` flag or removing it -- the caller can wrap with their own progress reporting.

File: vllm_omni/model_executor/stage_configs/moss_tts.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+      model_arch: MossTTSDelayModel
+      worker_type: generation
+      scheduler_cls: vllm_omni.core.sched.omni_generation_scheduler.OmniGenerationScheduler
+      enforce_eager: true
+      trust_remote_code: true
+      async_scheduling: false
+      enable_prefix_caching: false
+      engine_output_type: audio
+      gpu_memory_utilization: 0.1
+      distributed_executor_backend: "mp"
```
Comment (lishunyang12) [inline vllm_omni/model_executor/stage_configs/moss_tts.yaml:18]: `gpu_memory_utilization: 0.1` reserves only 10% of GPU for KV cache, which seems very low for an 8B model. The PR description mentions OOM on 4090 and needing A100. Was this chosen to leave room for the eager HF load? If the model were loaded through vLLM normal path this probably would not need to be so conservative.

File: vllm_omni/engine/arg_utils.py (hunk 4 above)
Comment (lishunyang12) [inline vllm_omni/engine/arg_utils.py:148]: Agreeing with @linyueqian -- this save/restore pattern for `trust_remote_code` is duplicated across both `create_model_config` methods and is tightly coupled to a single model arch string. If another model needs the same treatment you will end up with a growing list of arch checks here. Would a config-level flag (e.g. `prefer_local_config = True` on `MossTTSDelayConfig`) be cleaner than special-casing in the engine args?

File: vllm_omni/model_executor/models/moss_tts/moss_tts.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+
+        Args:
+            **kwargs: Multimodal input arguments
+
+        Returns:
+            None: TTS models don't use multimodal embeddings
+        """
+        return None
+
+    def load_weights(
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/moss_tts.py:202]: This records every param name as loaded without actually loading anything, since the real loading happened in `__init__` via HF `from_pretrained`. This is misleading to the framework -- vLLM trusts this method to report which params it actually consumed. If a param name changes upstream, this silently passes.

**Review by lishunyang12 — CHANGES_REQUESTED** [review]
Blocking on the transformers 5.x dependency and the HF eager-load bypassing vLLM weight loading, both already flagged by @linyueqian -- adding a few more things I noticed.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by Gaohan123** [comment]
@anurag12-webster Hello, any updates?

**Comment by anurag12-webster** [comment]
back on this will address the pending reviews shortly

**Comment by faisal-fida** [comment]
any updates?

**Comment by danich786** [comment]
Hi @anurag12-webster is the feature complete or you still working on solving issues?

**Comment by lishunyang12** [comment]
Closed as the model has been supported in https://github.com/vllm-project/vllm-omni/pull/3420. Thanks for your contribution.
